src/Stack/Setup.hs

```diff
@@ -2047,32 +2046,23 @@ downloadStackExe platforms0 archiveInfo destDir checkPath testExe = do
     handleTarball :: Path Abs File -> Bool -> T.Text -> IO ()
     handleTarball tmpFile isWindows url = do
         req <- fmap setGithubHeaders $ parseUrlThrow $ T.unpack url
-        withResponse req $ \res -> do
-            entries <- fmap (Tar.read . LBS.fromChunks)
-                     $ lazyConsume
-                     $ getResponseBody res .| ungzip
-            let loop Tar.Done = error $ concat
-                    [ "Stack executable "
-                    , show exeName
-                    , " not found in archive from "
-                    , T.unpack url
-                    ]
-                loop (Tar.Fail e) = throwM e
-                loop (Tar.Next e es)
-                    | Tar.entryPath e == exeName =
-                        case Tar.entryContent e of
-                            Tar.NormalFile lbs _ -> do
-                              ensureDir destDir
-                              LBS.writeFile (toFilePath tmpFile) lbs
-                            _ -> error $ concat
-                                [ "Invalid file type for tar entry named "
-                                , exeName
-                                , " downloaded from "
-                                , T.unpack url
-                                ]
-                    | otherwise = loop es
-            loop entries
+        result <- withResponse req $ \res ->
+            runConduitRes $ getResponseBody res
+                         .| ungzip
+                         .| CTar.untar extractExe
+                         .| CL.foldMap Any
+        unless (getAny result) $ error $ concat [ "Stack executable "
+                                                , show exeName
+                                                , " not found in archive from "
+                                                , T.unpack url
+                                                ]
       where
+        extractExe fi
+          | CTar.fileType fi == CTar.FTNormal && CTar.getFileInfoPath fi == exeName = do
+              ensureDir destDir
+              CB.sinkFile (toFilePath tmpFile)
+              yield True
+          | otherwise = yield False
```

Comment:
Add here a check and an error message when exec file is found but is of a wrong type, same as before:
```
                                [ "Invalid file type for tar entry named "	
                                , exeName	
                                , " downloaded from "	
                                , T.unpack url	
                                ]
```